File: CucumberJVM/Chapter15/15.7-StubDependencyPattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/NotificationsStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.models.Notification;
import com.wimp.app.specs.drivers.NotificationsApiDriver;
import com.wimp.app.specs.support.DomainDefaults;
import io.cucumber.java.en.*;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class NotificationsStepDefinitions {
    private final NotificationsApiDriver notificationsApiDriver;

    public NotificationsStepDefinitions(NotificationsApiDriver notificationsApiDriver) {
        this.notificationsApiDriver = notificationsApiDriver;
    }

    @Then("the customer should receive a notification about the delay")
    public void theCustomerShouldReceiveANotificationAboutTheDelay() throws Exception {
        List<Notification> notifications = notificationsApiDriver.getNotifications(DomainDefaults.CUSTOMER_NAME).execute();

        assertTrue(
            notifications.stream().anyMatch(n -> n.message().toLowerCase().contains("delayed")),
            "Expected a delay notification but none was found.");
    }
}

File: Reqnroll/Chapter15/15.7-StubDependencyPattern/Before/WIMP.Specs/StepDefinitions/NotificationsStepDefinitions.cs
using Reqnroll;

using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class NotificationsStepDefinitions(NotificationsApiDriver notificationsApiDriver)
{
    [Then("the customer should receive a notification about the delay")]
    public async Task ThenTheCustomerShouldReceiveANotificationAboutTheDelay()
    {
        var notifications = await notificationsApiDriver.GetNotifications(DomainDefaults.CustomerName).Execute();

        Assert.IsNotNull(notifications);
        Assert.IsTrue(notifications.Any(n =>
                n.Message.Contains("delayed", StringComparison.OrdinalIgnoreCase)),
            "Expected a delay notification but none was found.");
    }
}
